File: Cuarto/Laboratorio/Talleres/RobocodeLabo/src/Robotito2/EstrategiaOfensiva.java
package Robotito2;

import robocode.JuniorRobot;

public class EstrategiaOfensiva implements CombatStrategy {

    private int steps;

    @Override
    public void init(JuniorRobot robot) {
        robot.setColors(JuniorRobot.red, JuniorRobot.black, JuniorRobot.black,
                JuniorRobot.red, JuniorRobot.red);

        while (true){
            steps = (int)(Math.random() * 200 + 1);
            robot.ahead(steps);

            steps = (int)(Math.random() * 200 + 1);
            robot.back(steps);

            int inward = (robot.heading + 90) % 360;
            robot.turnGunTo((inward) % 360);
        }
    }

    private void discreteFire(JuniorRobot robot,double power){
        if (robot.gunReady) {
            robot.fire(power);
        }

    }

    @Override
    public void onTick(JuniorRobot robot) {

    }

    @Override
    public void onScannedRobot(JuniorRobot robot){
        robot.turnGunTo(robot.scannedAngle);
        // Ajustar potencia en función de la distancia
        double power = (robot.scannedDistance < 200) ? bulletMaxPower : bulletHalfPower;
        this.discreteFire(robot,power);


    }

    @Override
    public void onHitWall(JuniorRobot robot) {

    }

    @Override
    public void onHitRobot(JuniorRobot robot) {
        robot.turnGunTo(robot.hitRobotAngle);
        robot.fire(bulletMaxPower);

    }

    @Override
    public void onHitByBullet(JuniorRobot robot) {
        robot.turnGunTo(robot.hitByBulletAngle);
        this.discreteFire(robot,bulletHalfPower);

    }

}

File: Cuarto/Laboratorio/Talleres/RobocodeLabo/src/robotitoKotlin/EstrategiaOfensiva.kt
package robotitoKotlin

import robocode.JuniorRobot
import kotlin.random.Random

class EstrategiaOfensiva(robot: JuniorRobot): CombatStrategy(robot) {
    private var steps = Random.nextInt(1,200)


    //Bloque de inicializacion
    init {
        robot.setColors(JuniorRobot.red, JuniorRobot.black, JuniorRobot.black,
            JuniorRobot.red, JuniorRobot.red)
    }

    private fun discreteFire(power: Double){
        if(robot.gunReady) robot.fire(power)
    }

    override fun onTick(){
        robot.ahead(steps)

        steps = Random.nextInt(1,200)
        robot.back(steps)

        val inward = (robot.heading + 90) % 360
        robot.turnGunTo((inward) % 360)

    }

    override fun onScannedRobot() {
        robot.turnGunTo(robot.scannedAngle)

        val power = if (robot.scannedDistance > 200) this.bulletMaxPower else this.bulletHalfPower

        this.discreteFire(power)

    }

    override fun onHitWall() {

    }

    override fun onHitByBullet() {
        robot.turnGunTo(robot.hitByBulletAngle)
        this.discreteFire(this.bulletHalfPower)
    }


}